app/src/main/java/org/mozilla/fenix/settings/search/AddSearchEngineFragment.kt

```diff
@@ -0,0 +1,282 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.settings.search
+
+import android.content.res.Resources
+import android.graphics.drawable.BitmapDrawable
+import android.os.Bundle
+import androidx.fragment.app.Fragment
+import android.view.LayoutInflater
+import android.view.Menu
+import android.view.MenuInflater
+import android.view.MenuItem
+import android.view.View
+import android.view.ViewGroup
+import android.widget.CompoundButton
+import androidx.appcompat.app.AppCompatActivity
+import androidx.constraintlayout.widget.ConstraintLayout
+import androidx.lifecycle.lifecycleScope
+import androidx.navigation.fragment.findNavController
+import kotlinx.android.synthetic.main.custom_search_engine.*
+import kotlinx.android.synthetic.main.fragment_add_search_engine.*
+import kotlinx.android.synthetic.main.search_engine_radio_button.view.*
+import kotlinx.coroutines.Dispatchers.Main
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import mozilla.components.browser.search.SearchEngine
+import org.mozilla.fenix.R
+import org.mozilla.fenix.components.FenixSnackbar
+import org.mozilla.fenix.components.searchengine.CustomSearchEngineStore
+import org.mozilla.fenix.ext.components
+import org.mozilla.fenix.ext.increaseTapArea
+import org.mozilla.fenix.ext.requireComponents
+import org.mozilla.fenix.settings.SupportUtils
+import java.util.Locale
+
+@SuppressWarnings("LargeClass", "TooManyFunctions")
+class AddSearchEngineFragment : Fragment(), CompoundButton.OnCheckedChangeListener {
+    private var availableEngines: List<SearchEngine> = listOf()
+    private var selectedIndex: Int = -1
+    private val engineViews = mutableListOf<View>()
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setHasOptionsMenu(true)
+
+        availableEngines = runBlocking {
+            requireContext()
+                .components
+                .search
+                .provider
+                .uninstalledSearchEngines(requireContext())
+                .list
+        }
+
+        selectedIndex = if (availableEngines.isEmpty()) CUSTOM_INDEX else FIRST_INDEX
+    }
+
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? {
+        return inflater.inflate(R.layout.fragment_add_search_engine, container, false)
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        val layoutInflater = LayoutInflater.from(context)
+        val layoutParams = ViewGroup.LayoutParams(
+            ViewGroup.LayoutParams.MATCH_PARENT,
+            ViewGroup.LayoutParams.WRAP_CONTENT
+        )
+
+        val setupSearchEngineItem: (Int, SearchEngine) -> Unit = { index, engine ->
```

Comment:
Can a recycler view be used instead? I think it would better encasulate the logic of displaying these search engine options, especially if we eventually have a lot. That should also make it easier to have `availableEngines` loaded async.

Reply:
Yeah I think so. Though I don't think we will hit that limit any time soon. If we decide we like this we can port it to a recycler view.

Reply:
While we probably don't need it for perf I still feel like its easier to read the code as an adapter